black_it/samplers/particle_swarm.py

```diff
@@ -0,0 +1,233 @@
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+"""Implementation of the particle swarm sampler."""
+from typing import Optional, cast
+
+import numpy as np
+from numpy.typing import NDArray
+
+from black_it.samplers.base import BaseSampler
+from black_it.search_space import SearchSpace
+from black_it.utils.base import assert_, digitize_data, positive_float
+
+
+class ParticleSwarmSampler(
+    BaseSampler
+):  # pylint: disable=(too-many-instance-attributes)
+    """
+    Implementation of a particle swarm sampler.
+
+    This sampler implements the particle swarm sampling method commonly used in particle swarm optimization (PSO),
+    introduced in:
+
+        Eberhart, Russell, and James Kennedy. "A new optimizer using particle swarm theory."
+          MHS'95. Proceedings of the sixth international symposium on micro machine and human science. IEEE, 1995.
+
+    In a particle swarm optimizer, there is a set of particles that are "evolved" by cooperation and competition
+    among the individuals themselves through generations. Each particle adjusts its flying  according to its own
+    flying experience and its companions’ flying experience. Each particle, in fact, represents a potential solution
+    to a problem. Each particle is treated as a point in a D-dimensional space.  The ith particle is represented as
+    Xi = (x_{i1},...,x_{iD}). The best previous position (the position giving the best fitness value) of any particle
+    is recorded and represented as Pi = (p_{i1},...,p_{iD}). The index of the best particle among all the particles
+    in the population is represented by the symbol g. The rate of the position change (velocity) for particle i is
+    represented as Vi = (v_{i1},...,v_{iD}). The particles are manipulated according to the following equation:
+
+        v_{id} = (ω * v_{id}) + (c1 * r1 * (p_{id} - x_{id})) + (c2 * r2 * (p_{gd} - x_{id})
+        x_{id} = x_{id} + v_{id}
+
+    Where:
+        - ω is the inertia weight  to control the influence of the previous velocity;
+        - c1 and c2 are positive values that represent the acceleration constants;
+        - r1 and r2 are two random numbers uniformly distributed in the range of (0, 1).
+
+    This sampling method has to be used alone in Black-it Calibrator, because existing_losses and existing_points
+    arguments of sample_batch are assumed to be computed by this sampler.
```

Comment:
in fact, the docstring is not quite correct (my bad). `_previous_batch_index_start` will necessarily point to the index of `existing_points`/`existing_losses` at the moment of the previous call to `sample_batch`. Any other existing batch is ignored. Therefore, atm, the PS sampler works as if it was alone.

We could switch the behaviour by considering the global loss as the true global loss of the calibration task, and not just the global loss within the PS sampler.

Reply:
I kind of like this idea actually... It would allow different samplers to benefit from each other, and this is one of the nice features of Black-it I think